### PR TITLE
CI: Add cirrus-ci to test linux_aarch64

### PR DESCRIPTION
**Pull request by andyfaff**
Addresses a large proportion of #22640. This PR adds cibuildwheel (build+test) for linux_aarch64 on cirrus-ci.

### CODE REVIEW

**Comment by mattip** [comment]
Are the 3.9 and 3.10 builds joined into one task on purpose?

**Comment by andyfaff** [comment]
I guess to save on firing another VM up. They can all be individual jobs if you want

**Comment by mattip** [comment]
Maybe a comment to say why two together and not 3 separate or 3 together (it would push the buld/test over the time limit?)

**Comment by andyfaff** [comment]
I split into separate runs, probably easier to follow.

**Comment by mattip** [comment]
I went to log in to cirrus CI via my github login to create the tokens, but got this screen. As a numpy/numpy admin I am a little hesitant to let them "act on my behalf" which, as I understand it, could mean any actions I can do in the repo.
Screenshot from 2022-11-26 19-03-34

**Comment by andyfaff** [comment]
@mattip did it give more detail under *learn more*? If I look at the installed app on my personal account it has the following permissions:
- Read access to files located at .cirrus.yml
- Read access to code, metadata, and pull requests
-  Read and write access to checks and commit statuses
-  Write access to attach content to the following external domain:
cirrus-ci.com

These seem reasonable to me. In any case I think you need to be adding encrypted variables at https://cirrus-ci.com/github/numpy/numpy. There should be a gear/settings icon at the top RHS. You press that and you should come up with a settings page where you can enter *secured variables*. That's where you put the anaconda token you generated. It spits back a string that needs to be entered into the PR [here](https://github.com/numpy/numpy/pull/22650/files#diff-ec0a015d68b2388be2fe17d4f081fbd5fd4741e7835a507f23fc7debf6bcaff1R62)

Screenshot 2022-11-27 at 11 26 57

@rgommers did this for scipy.

File: tools/ci/cirrus_general.yml [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+  depends_on:
+    - cirrus_wheels_linux_aarch64
+  compute_engine_instance:
+    image_project: cirrus-images
+    image: family/docker-builder
+    platform: linux
+
+#  env:
+#    NUMPY_STAGING_UPLOAD_TOKEN: ENCRYPTED[]
+#    NUMPY_NIGHTLY_UPLOAD_TOKEN: ENCRYPTED[]
```
Comment (mattip) [inline tools/ci/cirrus_general.yml:63]:
```suggestion
  env:
    NUMPY_STAGING_UPLOAD_TOKEN: ENCRYPTED[!5a69522ae0c2af9edb2bc1cdfeaca6292fb3666d9ecd82dca0615921834a6ce3b702352835d8bde4ea2a9ed5ef8424ac!]
    NUMPY_NIGHTLY_UPLOAD_TOKEN: ENCRYPTED[!196422e6c3419a3b1d79815e1026094a215cb0f346fe34ed0f9d3ca1c19339df7398d04556491b1e0420fc1fe3713289!]```

**Comment by mattip** [comment]
I added encrypted tokens from the appropriate `https://anaconda.org/*/settings/access` location

File: tools/ci/cirrus_general.yml [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    fi
+
+    # The name of the zip file is derived from the `wheels_artifact` line.
+    # If you change the artifact line to `myfile_artifact` then it would be
+    # called myfile.zip
+    
+    curl https://api.cirrus-ci.com/v1/artifact/build/$CIRRUS_BUILD_ID/wheels.zip --output wheels.zip
+    unzip wheels.zip
+    
+    # TODO upload wheels
```
Comment (mattip) [inline tools/ci/cirrus_general.yml:86]:
```suggestion
```

**Comment by mattip** [comment]
Thanks @andyfaff

**Comment by mattip** [comment]
@andyfaff do you know how to set the `api` environment variable when manually triggering a build via the + button [on the gui](https://cirrus-ci.com/github/numpy/numpy)

Screenshot from 2022-11-28 15-23-49

The build is [called "API build"](https://cirrus-ci.com/task/6054489367511040), 

Screenshot from 2022-11-28 15-24-11



I guess we need to test `TRAVIS_EVENT_TYPE` to trigger this line https://github.com/numpy/numpy/blob/bba74740e5236735f6ec5013d8d778145b758e89/tools/wheels/upload_wheels.sh#L12

**Comment by andyfaff** [comment]
@mattip 

TLDR: 
- strongly advise removing the [`set_travis_vars` function call](https://github.com/numpy/numpy/blob/main/tools/ci/cirrus_general.yml#L88)
- add the bash code at the end to detect whether a manual run has started and set `IS_SCHEDULE_DISPATCH="true"`.

You asked a question:

> do you know how to set the api environment variable when manually triggering a build via the + button [on the gui](https://cirrus-ci.com/github/numpy/numpy)?

But I can't seem to find it in this PR any more. This PR **intentionally didn't use `set_travis_vars`** from [`tools/wheels/upload_wheels.sh`](https://github.com/numpy/numpy/blob/main/tools/ci/cirrus_general.yml#L88). There is no need to call that bash function because none of those environment variables used in that function (e.g. `TRAVIS_EVENT_TYPE`) will be set because it's not a TravisCI run. In fact calling that function is probably detrimental to the upload because I think the function will set `IS_SCHEDULE_DISPATCH="false"`, `IS_PUSH="false"` because those environment variables aren't present. This will undo work done [here](https://github.com/numpy/numpy/blob/main/tools/ci/cirrus_general.yml#L71) that goes to the effort of setting `IS_PUSH` and `IS_SCHEDULE_DISPATCH` in the cirrus script. I would advise removing the `set_travis_vars` call.

I have just been experimenting with a manual api build on my scipy fork. I created a manual build in the manner you suggested, printing out environment variables from the run. Here is the comparison from a normal push to the manual press. I have removed the environment variables that are the same between the two runs.

[Normal (i.e. a PR/push/etc)](https://cirrus-ci.com/task/4786888893005824)
```
CIRRUS_CHANGE_TITLE=TST
CIRRUS_CHANGE_MESSAGE=TST
CIRRUS_COMMIT_MESSAGE=TST
CIRRUS_BUILD_SOURCE=github
```

[Manual build (on main)](https://api.cirrus-ci.com/v1/task/5640583100760064/logs/test.log):
```
CIRRUS_CHANGE_TITLE=API build for null
CIRRUS_CHANGE_MESSAGE=API build for null
CIRRUS_COMMIT_MESSAGE=API build for null
CIRRUS_BUILD_SOURCE=api
```

Thus, to test for a manual build, and therefore set `IS_SCHEDULE_DISPATCH="true"`, it may be sufficient to add the following to cirrus_general.yml:

```
if [[ $CIRRUS_BUILD_SOURCE=="api" && $CIRRUS_COMMIT_MESSAGE=="API build for null" ]]; then
    export IS_SCHEDULE_DISPATCH="true"
fi
```
